database2/ffldb/db.go

```diff
@@ -0,0 +1,2087 @@
+// hash.  It will return ErrBlockNotFound if there is no entry.
+func (tx *transaction) fetchBlockRow(hash *chainhash.Hash) ([]byte, error) {
+	blockRow := tx.blockIdxBucket.Get(hash[:])
+	if blockRow == nil {
+		str := fmt.Sprintf("block %s does not exist", hash)
+		return nil, makeDbErr(database.ErrBlockNotFound, str, nil)
+	}
+
+	return blockRow, nil
+}
+
+// FetchBlockHeader returns the raw serialized bytes for the block header
+// identified by the given hash.  The raw bytes are in the format returned by
+// Serialize on a wire.BlockHeader.
+//
+// Returns the following errors as required by the interface contract:
+//   - ErrBlockNotFound if the requested block hash does not exist
+//   - ErrTxClosed if the transaction has already been closed
+//   - ErrCorruption if the database has somehow become corrupted
+//
+// NOTE: The data returned by this function is only valid during a
+// database transaction.  Attempting to access it after a transaction
+// has ended results in undefined behavior.  This constraint prevents
+// additional data copies and allows support for memory-mapped database
+// implementations.
+//
+// This function is part of the database.Tx interface implementation.
+func (tx *transaction) FetchBlockHeader(hash *chainhash.Hash) ([]byte, error) {
+	// Ensure transaction state is valid.
+	if err := tx.checkClosed(); err != nil {
+		return nil, err
+	}
+
+	// When the block is pending to be written on commit return the bytes
+	// from there.
+	if idx, exists := tx.pendingBlocks[*hash]; exists {
+		blockBytes := tx.pendingBlockData[idx].bytes
+		return blockBytes[0:blockHdrSize:blockHdrSize], nil
+	}
+
+	// Fetch the block index row and slice off the header.  Notice the use
+	// of the cap on the subslice to prevent the caller from accidentally
+	// appending into the db data.
+	blockRow, err := tx.fetchBlockRow(hash)
+	if err != nil {
+		return nil, err
+	}
+	endOffset := blockLocSize + blockHdrSize
+	return blockRow[blockLocSize:endOffset:endOffset], nil
+}
+
+// FetchBlockHeaders returns the raw serialized bytes for the block headers
+// identified by the given hashes.  The raw bytes are in the format returned by
+// Serialize on a wire.BlockHeader.
+//
+// Returns the following errors as required by the interface contract:
+//   - ErrBlockNotFound if the any of the requested block hashes do not exist
+//   - ErrTxClosed if the transaction has already been closed
+//   - ErrCorruption if the database has somehow become corrupted
+//
+// NOTE: The data returned by this function is only valid during a database
+// transaction.  Attempting to access it after a transaction has ended results
+// in undefined behavior.  This constraint prevents additional data copies and
+// allows support for memory-mapped database implementations.
+//
+// This function is part of the database.Tx interface implementation.
+func (tx *transaction) FetchBlockHeaders(hashes []chainhash.Hash) ([][]byte, error) {
```

Comment:
FetchBlockHeaders takes a slice of values, but FetchBlockHeader takes a pointer to a value?